p2p/src/channels/p2p_channels_reducer.rs

```diff
@@ -1,30 +1,210 @@
+};
+use openmina_core::{block::BlockWithHash, error, Substate};
+use redux::{ActionWithMeta, Dispatcher};
 
 impl P2pChannelsState {
-    pub fn reducer(&mut self, action: P2pChannelsActionWithMetaRef<'_>, is_libp2p: bool) {
+    pub fn reducer<Action, State>(
+        state_context: Substate<Action, State, P2pState>,
+        action: ActionWithMeta<&P2pChannelsAction>,
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
         let (action, meta) = action.split();
+
         match action {
-            P2pChannelsAction::MessageReceived(_) => {}
+            P2pChannelsAction::MessageReceived(action) => {
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                Self::dispatch_message(meta.with_action(action), dispatcher, state)
+            }
             P2pChannelsAction::BestTip(action) => {
-                self.best_tip.reducer(meta.with_action(action), is_libp2p);
+                P2pChannelsBestTipState::reducer(state_context, meta.with_action(action))
             }
             P2pChannelsAction::Transaction(action) => {
-                self.transaction.reducer(meta.with_action(action));
+                P2pChannelsTransactionState::reducer(state_context, meta.with_action(action))
             }
             P2pChannelsAction::Snark(action) => {
-                self.snark.reducer(meta.with_action(action));
+                P2pChannelsSnarkState::reducer(state_context, meta.with_action(action))
             }
             P2pChannelsAction::SnarkJobCommitment(action) => {
-                self.snark_job_commitment.reducer(meta.with_action(action));
+                P2pChannelsSnarkJobCommitmentState::reducer(state_context, meta.with_action(action))
             }
             P2pChannelsAction::Rpc(action) => {
-                self.rpc
-                    .reducer(meta.with_action(action), &mut self.next_local_rpc_id);
+                P2pChannelsRpcState::reducer(state_context, meta.with_action(action))
             }
             P2pChannelsAction::StreamingRpc(action) => {
-                self.streaming_rpc
-                    .reducer(meta.with_action(action), &mut self.next_local_rpc_id);
+                P2pChannelsStreamingRpcState::reducer(state_context, meta.with_action(action))
             }
         }
     }
+
+    fn dispatch_message<Action, State>(
+        action: ActionWithMeta<&P2pChannelsMessageReceivedAction>,
+        dispatcher: &mut Dispatcher<Action, State>,
+        state: &State,
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
+        let (action, meta) = action.split();
+        let time = meta.time();
+
+        let peer_id = action.peer_id;
+        let chain_id = action.message.channel_id();
+
+        let mut is_enabled = |action: Action| dispatcher.push_if_enabled(action, state, time);
```

Comment:
I have to look at it more thoroughly, but be careful with this. `push_is_enabled` will return a result based on the state at the time the action was queued, which may not be the same as the state when the action is finally dispatched.